new-packages/Core/src/Ranges/Progressions.candy

```diff
@@ -0,0 +1,114 @@
+use ...Bool
+use ...Collections
+use ...Int
+use ...Maybe
+use ...Operators
+
+public trait IntProgression: Equals {
+  fun startBound(): Maybe[Int]
+  fun endBound(): Maybe[Int]
+  fun stepSize(): Int
+
+  fun contains(item: Int): Bool {
+    let startMatches = this.startBound()
+      .map[Bool]({ it <= item & (item - it) % this.stepSize() == 0 })
+      .else({ true() })
+    let endMatches = this.endBound()
+      .map[Bool]({ item <= it & (it - item) % this.stepSize() == 0 })
+      .else({ true() })
+    startMatches & endMatches
+  }
+}
+
+
+public type BoundedIntProgression = (start: Int, end: Int, step: Int)
+# A `Progression` where `this.start` is inclusive and `this.end` is exclusive.
+#
+# Invariants:
+# * `this step != 0`
+# * `(this step > 0) then { this end > this start } else { this end < this start }`
```

Comment:
How is an empty progression represented so that these two invariants are true?

Reply:
Fixed the invariant: `this end` is equal to `this start` to represent an empty progression.